env-var-docs/parser-package/src/env_var_docs/parser.py

```diff
@@ -0,0 +1,522 @@
+
+
+VisitFn = typing.Callable[[NodeInfo], None]
+"""A function that takes a NodeInfo and does something with it."""
+
+
+def visit(file: typing.TextIO, visit_fn: VisitFn) -> list[NodeParseError]:
+    """Parses an environment variable documentation file. The file must be
+    valid JSON and be a single object. See README.md for the expected
+    structure.
+
+    The file is completely parsed before any visit_fn is run. If there are
+    parsing errors, no visit_fn is run and the errors are returned.
+
+    If the file has no parsing errors, visit_fn is called on each node in
+    preorder traversal (visit node then its children) order.
+    """
+    file.seek(0)  # Ensure we are reading from the beginning of the file.
+
+    json_root = "file"
+    try:
+        docs = json.load(file)
+    except json.JSONDecodeError as e:
+        return [
+            NodeParseError(
+                json_root,
+                [ParseError(json_root, f"file is not valid JSON: {e}")], [])
+        ]
+    if not isinstance(docs, dict):
+        return [
+            NodeParseError(
+                json_root,
+                [ParseError(json_root, f"file is not a single JSON object")],
+                [])
+        ]
+
+    nodes_to_visit: list[NodeInfo] = []
+    parsing_errors: list[NodeParseError] = []
+
+    # The file is an implicit root node. Recursively descend into each top-level field.
+    for key, value in docs.items():
+        _recursively_parse(
+            1, json_root, key, value, nodes_to_visit, parsing_errors)
+
+    if len(parsing_errors) != 0:
+        return parsing_errors
+
+    for node in nodes_to_visit:
+        visit_fn(node)
+
+    return []
+
+
+def _recursively_parse(
+        level: int, parent_path: str, key: str, value: UnparsedJSON,
+        nodes_to_visit: list[NodeInfo], parsing_errors: list[NodeParseError]):
+    """Recursively visits a node and its children, in preorder traversal order (visit root then children)."""
+    path = _path(parent_path, key)
+
+    # Is this a group?
+    group, group_errors = _parse_group(path, value)
+    if len(group_errors) == 0:
+        assert group is not None  # appease mypy
+        nodes_to_visit.append(
+            NodeInfo(
+                level=level, json_path=parent_path, name=key, details=group))
+
+        # Recurse into group members.
+        for sub_key, sub_value in group.members.items():
+            _recursively_parse(
+                level + 1, path, sub_key, sub_value, nodes_to_visit,
+                parsing_errors)
+        return
+
+    # Is this a variable?
+    var, var_errors = _parse_variable(path, value)
+    if len(var_errors) == 0:
+        assert var is not None  # appease mypy
+        nodes_to_visit.append(
+            NodeInfo(level=level, json_path=parent_path, name=key, details=var))
+        return
+
+    # details is neither a group nor a variable.
+    parsing_errors.append(NodeParseError(path, group_errors, var_errors))
```

Comment:
**thought (non-blocking):** looks like if a node is malformed we present the errors for it not being a var, ignoring that it could be a malformed group. That seems fine but want to double check that's the intention.

Reply:
The `NodeParseError` type wraps the `ParseError` type so that callers of `visit` can separate the parse errors resulting from attempting to parse the object as a `Group` and the errors resulting from attempting to parse as a `Variable`.

The one place this abstraction breaks down is at the beginning of `visit` where it checks if the file is valid JSON and that it contains a single object.  In these cases I return a `NodeParseError` with the file-level error reported in the `group_errors` field.  This is not strictly correct because the file root is neither a `Group` nor a `Variable` but it seemed good enough rather than further complicating `NodeParseError` with a `file_errors` field.

Reply:
Sounds good, thanks for the explanation